Documentation overhaul - Fixing formatting issues, autodoc implementation, and added variables for versions.

wes-otf opened the pull request:
<!--
Thanks for contributing to Hypha!

Please ensure your contributions pass all necessary linting/testing and that the appropriate documentation has been updated.
-->

## Description
<!--
Describe briefly what your pull request changes. If this is resoving an issue, please specify below via "Fixes #<Github Issue ID>"
-->
Closes #3639. This is a bit of an overhaul of the docs to make them more easy to utilize. This changes include:
- Removal of overhanging public aspects of the documentation
- Reformatting of terminal commands to allow for easy copy & pasting
- Updating points of contact
- Screenshot Updates
- Automated code API documentation. [Google's docstring style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings) should be used.
- Added variables for things like software versions & packages.
  - Unless python & node version are manually specified, when docs are built these will be pulled from [`.nvmrc`](https://github.com/HyphaApp/hypha/blob/main/.nvmrc) & [`.python-version`](https://github.com/HyphaApp/hypha/blob/main/.python-version)
- Consolidated the production and development guides into one location (both under `Setup` rather than one under `Getting Started` and one under `Setup`), and updated them to be similar to one another.
- Added different steps for getting Hypha development deployed under multiple OSs

## Screenshots

*API Documentation*
<img width="1390" alt="Screenshot 2024-01-23 at 10 48 58" src="https://github.com/HyphaApp/hypha/assets/145372368/7796ee8d-e73d-407a-95da-60a180a8aca0">

## Test Steps
<!-- 
If step does not require manual testing, skip/remove this section.

Give a brief overview of the steps required for a user/dev to test this contribution. Important things to include:
 - Required user roles for where neccesary (ie. "As a Staff Admin...")
 - Clear & validatable expected results (ie. "Confirm the submit button is now not clickable")
 - Langauge that can be understood by non-technical testers if being tested by users
-->

This would just require developer verification that docs are being built and generated as expected. Specifically automatic versioning and API documentation is being generated.

wes-otf committed:
Updated main urls.py to include both the users app public_urlpatterns & urlpatterns

wes-otf committed:
Added more detailed installation instructions, fixed formatting issues, and added variables to track software versions.

wes-otf committed:
Formatting fixes

wes-otf committed:
Reorganized development setup to be under deployment, more formatting fixes and changes after full Ubuntu test

wes-otf committed:
Package & macro feature updates

wes-otf committed:
Reaaranged deployment instructions

wes-otf committed:
Removed outdated copy of stand-alone.md

wes-otf committed:
Removed accidental urls.py edit

wes-otf committed:
Merged with main

wes-otf committed:
Merge branch 'main' into enhancement/doc-improvements

wes-otf committed:
Doc restructuring

wes-otf committed:
Merged with main

wes-otf committed:
Formatting changes, removed more mentions of the public site, and updated vulnerability disclosure info

wes-otf committed:
Formatting changes, adding notification section

wes-otf committed:
Updated login guide to include passwordless workflow & updated screenshots

wes-otf committed:
README updates

wes-otf committed:
Merged with main

wes-otf committed:
Initial implementation of docstring automation

wes-otf committed:
Formatting fixes, fixed some existing docstrings

wes-otf committed:
Merge branch 'main' into enhancement/doc-improvements

wes-otf committed:
Moved API Reference to under references dir, updates to stand alone docs

wes-otf committed:
Updated comments in gen_ref_pages script

wes-otf committed:
Add automation for Python & Node versions based off of .python-version & .nvmrc

wes-otf committed:
Docstring comment fix

wes-otf committed:
Final additions, added deployment page to getting started & link updates

wes-otf committed:
Renamed service.py due to the addition of __init__.py

wes-otf commented:
[0522725](https://github.com/HyphaApp/hypha/pull/3734/commits/052272541f13ce8ba9d31e7218a69091a02ebcd0]) was made on accident and was removed in a following commit

theskumar commented on mkdocs.yml:
```diff
@@ -40,8 +40,33 @@ theme:
   - search
+  - macros:
+      module_name: docs/scripts/doc_macros
+  - mkdocstrings:
```
> Automated code API documentation. [Google's docstring style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings) should be used.

This has significantly increased the `mkdocs serve` time to refresh or just being useable? Would be possible to have this as optional while writing docs. Or provide an alternative to preview docs quickly?

theskumar replied:
Looking deeper this seems to be easy doable if the `gen_ref_pages.py` uses a environment variable based flag to generate the initial reference files. Then enable that during build only by default. If someone want to enable during `mkdocs serve` it should be posssible too. 

```shell
GEN_REF_PAGES=True mkdocs build        # Generate reference docs, use in CI/CD
```

```shell
mkdocs serve        # skip reference docs, to be used locally / `Makefile` 
```

theskumar replied:
Alternatively, (better option) have `enabled` flag added to the `mkdocsstrings` config 

```yaml
  - mkdocstrings:
      enabled: !ENV [GEN_REF_PAGES, false]
      ...
```

wes-otf replied:
I also added an env check to gen_ref_files as that was still running w/o mkdocstrings being enabled.

theskumar commented on mkdocs.yml:
```diff
@@ -40,8 +40,33 @@ theme:
+  - gen-files:
+      scripts:
+        - docs/scripts/gen_ref_pages.py
+  - section-index
```
This is causing the section get hidden in the navigation section itself, unless there is going to be `index.md` in each of the folder with some over about the section, I believe will cause bit of confusion to readers. 

![image](https://github.com/HyphaApp/hypha/assets/236356/4f8bb4be-e91f-4a65-979b-8ba609a397f8)

theskumar replied:
Without `section-index` plugin. 

![Screenshot 2024-01-25 at 6  19 21@2x](https://github.com/HyphaApp/hypha/assets/236356/00d59880-4379-4b1b-a846-167070961266)

theskumar replied:
theme feature `navigation.indexes` (already enabled) probably serves the same purpose, but doesn't link the section header, if the `index.md` is missing. 

https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#section-index-pages

wes-otf replied:
Yeah you're totally right, I didn't even notice `section-index` was duplicating the `navigation.indexes` functionality. I tossed `section-index` and all seems to work as expected from what I've seen!

theskumar commented on requirements-docs.txt:
```diff
@@ -1,3 +1,7 @@
 mkdocs-material==9.4.6
```
Upgrade to `9.5.5`? it seems to have some nices fixes compared to this version.

https://squidfunk.github.io/mkdocs-material/changelog/

theskumar commented on requirements-docs.txt:
```diff
@@ -1,3 +1,7 @@
+mkdocs-macros-plugin
+mkdocstrings[python]
+mkdocs-gen-files
+mkdocs-section-index
```
I would suggest freezing the dependencies so we don't get accidentally broken docs.

theskumar commented on docs/scripts/gen_ref_pages.py:
```diff
@@ -0,0 +1,40 @@
+            identifier = f"::: hypha.{'.'.join(parts)}"
+            fd.write(identifier)
+
+    mkdocs_gen_files.set_edit_path(full_doc_path, path)
```
Faster/concurrent version, also removes "migrations" folder from being processed.

```suggestion
import concurrent.futures
from pathlib import Path

import mkdocs_gen_files

# The code source directory
src = Path(__file__).parent.parent.parent / "hypha"
skip_folders = ["public", "tests", "migrations"]

def process_path(path):
    module_path = path.relative_to(src).with_suffix("")
    doc_path = path.relative_to(src).with_suffix(".md")
    full_doc_path = Path("references/API", doc_path)

    parts = tuple(module_path.parts)

    # Don't document any code in public/ or tests/ directories
    for skip_folder in skip_folders:
        if skip_folder in parts:
            return

    if parts[-1] == "__init__":
        parts = parts[:-1]
        doc_path = doc_path.with_name("index.md")
        full_doc_path = full_doc_path.with_name("index.md")
    elif parts[-1] in ["__main__"]:
        return

    # Create the required files on build with mkdocs_gen_files, see:
    # https://oprypin.github.io/mkdocs-gen_files/
    if len(parts) > 0:
        with mkdocs_gen_files.open(full_doc_path, "w") as fd:
            identifier = f"::: hypha.{'.'.join(parts)}"
            fd.write(identifier)

    mkdocs_gen_files.set_edit_path(full_doc_path, path)

def generate_pages():
    with concurrent.futures.ThreadPoolExecutor() as executor:
        list(executor.map(process_path, sorted(src.rglob("*.py"))))

generate_pages()
```

wes-otf replied:
This definitely makes it a lot snappier! Great suggestion.

theskumar commented on docs/setup/deployment/production/stand-alone.md:
````diff
@@ -62,14 +147,18 @@ To begin with, set the `export SECURE_SSL_REDIRECT=false` to prevent SSL redirec
-* `python manage.py createsuperuser`
-* `python manage.py wagtailsiteupdate server.domain apply.server.domain 80`
-* `python manage.py runserver` \(runs development server at [http://127.0.0.1:8000](http://127.0.0.1:8000)\)
+```console
````
`console` stands for bash session. I believe the intension here is to have these as copy/paste shell script. Use `shell` marker instead?

theskumar replied:
![Screenshot 2024-01-25 at 7  57 26@2x](https://github.com/HyphaApp/hypha/assets/236356/bf39be66-d3cb-4fa8-b6bc-c1f24e11fa02)

wes-otf commented:
@theskumar This feedback is fantastic, thanks for taking the time to go over it all! I'll be handling/replying to all of these between today & tomorrow.

wes-otf committed:
Add concurrency to API page doc generation

Co-authored-by: Saurabh Kumar <theskumar@users.noreply.github.com>

wes-otf committed:
Froze deps, removed section-index, final changes to gen_ref_pages, and swapped `console` lexer for `shell`.

wes-otf commented on .github/workflows/deploy-docs.yml:
```diff
@@ -24,6 +24,7 @@ jobs:
       - name: Deploy docs
         uses: mhausenblas/mkdocs-deploy-gh-pages@master
         env:
+          GEN_REF_PAGES: true
```
@frjo @theskumar I believe this is the right place for this? Was between this & putting in one level up under the ubuntu-latest `env`.

theskumar replied:
I believe either would work. Feels right to me.

theskumar reviewed (approved):
💯

wes-otf committed:
Resolving merge conflicts & domain fixes

frjo commented:
Excellent work here!

I was wondering around the change `apply.server.domain` to `server.domain` without a subdomain.

While Hypha no longer have a public site I think most implementors will install Hypha on a subdomain. Root domain (and www) most likely point to their main web site.

I therefore think it is most helpful to show all examples with a subdomain.

Ideas what the best example subdomain to use is? Is it `apply.` or `hypha.` or something else?

wes-otf commented:
yeah that's a great point - probably in the config commands too then right? I think `apply.` is a good stand in. I'll revert now!

wes-otf committed:
Added the subdomain back to production and added note explaining why

wes-otf commented:
So I updated only the production side of things - to me it doesn't make much sense to have the subdomain be for local development too, but let me know your thoughts! I also added a note explaining why this is:
<img width="1160" alt="Screenshot 2024-02-06 at 16 20 10" src="https://github.com/HyphaApp/hypha/assets/145372368/19cad5dc-0b6a-458d-b08a-c6637e481671">

wes-otf committed:
Typo fix

frjo commented:
@wes-otf  This is ready to merge in I think. Anything you want to add first?

wes-otf commented:
@frjo I'm all set! I think it's ready to go.
